components/fonts/platform/macos/core_text_font_cache.rs

```diff
@@ -25,33 +25,78 @@ use crate::system_font_service::FontIdentifier;
 /// cache.
 pub(crate) struct CoreTextFontCache(OnceLock<RwLock<HashMap<FontIdentifier, CachedCTFont>>>);
 
+/// A cache of `VariationAxisInformation` to avoid having to read this information from
+/// a font everytime one is constructed.
+pub(crate) struct VariationAxisInformationCache(
+    OnceLock<RwLock<HashMap<FontIdentifier, Option<Vec<VariationAxisInformation>>>>>,
+);
+
 /// The global [`CoreTextFontCache`].
 static CACHE: CoreTextFontCache = CoreTextFontCache(OnceLock::new());
 
+/// The global [`VariationAxisInformationCache`] cache.
+static VARIATION_AXIS_INFORMATION_CACHE: VariationAxisInformationCache =
+    VariationAxisInformationCache(OnceLock::new());
+
 /// A [`HashMap`] of cached [`CTFont`] for a single [`FontIdentifier`]. There is one [`CTFont`]
 /// for each cached font size.
-type CachedCTFont = HashMap<Au, CTFont>;
+type CachedCTFont = HashMap<CoreTextFontCacheKey, PlatformFont>;
+
+#[derive(Eq, Hash, PartialEq)]
+struct CoreTextFontCacheKey {
+    size: Au,
+    variations: Vec<FontVariation>,
+}
 
 impl CoreTextFontCache {
     pub(crate) fn core_text_font(
         font_identifier: FontIdentifier,
         data: Option<&FontData>,
         pt_size: f64,
+        variations: &[FontVariation],
     ) -> Option<PlatformFont> {
         //// If you pass a zero font size to one of the Core Text APIs, it'll replace it with
         //// 12.0. We don't want that! (Issue #10492.)
         let clamped_pt_size = pt_size.max(0.01);
         let au_size = Au::from_f64_px(clamped_pt_size);
 
+        let key = CoreTextFontCacheKey {
+            size: au_size,
+            variations: variations.to_owned(),
+        };
+
         let cache = CACHE.0.get_or_init(Default::default);
         {
             let cache = cache.read();
-            if let Some(core_text_font) = cache
+            if let Some(platform_font) = cache
                 .get(&font_identifier)
-                .and_then(|identifier_cache| identifier_cache.get(&au_size))
+                .and_then(|identifier_cache| identifier_cache.get(&key))
             {
-                return Some(PlatformFont::new_with_ctfont(core_text_font.clone()));
+                return Some(platform_font.clone());
+            }
+        }
+
+        if !key.variations.is_empty() {
+            let core_text_font_no_variations =
+                Self::core_text_font(font_identifier.clone(), data, clamped_pt_size, &[])?;
```

Comment:
We could save some locking here by acquiring the write lock before the call and replacing it with a manual lookup, but that seems like premature optimization...